### ISSUE TITLE
yolov11-cls, val

### ISSUE THREAD

**Issue by liusse** [ultralytics/ultralytics#22343]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi！
I used a classification model to classify data, and when validating the data using a validation set, I want to obtain detailed information about misclassified data, such as their names. It seems that the existing parameters cannot achieve this functionality. Is there a simple and feasible way to implement this?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @liusse, thank you for your interest in Ultralytics 🚀! This is an automated response to help get you started—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., YOLO11-cls validation outputs), please share:
-1- Exact command or Python snippet you used for validation
-2- ultralytics and PyTorch versions, OS, and CUDA details
-3- A short sample of your dataset structure and a few example images
-4- Validation logs/console output showing the results you’re seeing
And please verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). This context will help us guide you on extracting the specific misclassification details you’re looking for 🙏

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8** to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- Amazon Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Thanks! You can log misclassified samples (with file path and true/pred labels) by using a tiny custom `ClassificationDataset` + `ClassificationValidator` that adds `im_file` to the batch and writes a CSV during `val()`; example below.

```python
import csv
from pathlib import Path
import torch
from ultralytics import YOLO
from ultralytics.data.dataset import ClassificationDataset
from ultralytics.models.yolo.classify import ClassificationValidator

# Include file paths in classification batches
class DatasetWithPaths(ClassificationDataset):
    def __getitem__(self, i: int):
        d = super().__getitem__(i)          # returns {'img': tensor, 'cls': int}
        d["im_file"] = self.samples[i][0]   # add image file path
        return d

# Record misclassifications to CSV
class MisclsValidator(ClassificationValidator):
    def build_dataset(self, img_path: str):
        return DatasetWithPaths(root=img_path, args=self.args, augment=False, prefix=self.args.split)

    def init_metrics(self, model):
        super().init_metrics(model)
        self.mis = []

    def update_metrics(self, preds: torch.Tensor, batch: dict):
        super().update_metrics(preds, batch)                    # keep top1/top5 + confusion matrix
        top1 = preds.argmax(1)
        probs = preds.softmax(1)
        for i, (t, p) in enumerate(zip(batch["cls"].view(-1), top1)):
            if p != t:
                self.mis.append((batch["im_file"][i], int(t), int(p), float(probs[i, p])))

    def finalize_metrics(self):
        super().finalize_metrics()
        if self.mis:
            out = self.save_dir / "misclassified.csv"
            with open(out, "w", newline="") as f:
                w = csv.writer(f)
                w.writerow(["file", "true", "pred", "conf", "true_name", "pred_name"])
                for fpath, t, p, conf in self.mis:
                    w.writerow([fpath, t, p, conf, self.names[t], self.names[p]])

# Run validation with the custom validator
model = YOLO("yolo11n-cls.pt")
metrics = model.val(data="path/to/your/val", validator=MisclsValidator)
```

Notes and references: `ClassificationDataset.__getitem__` returns only `{'img','cls'}`, so we add `im_file` ourselves; passing a custom validator/dataset is supported as shown in the Image Classification docs; and you can also export the confusion matrix via `results.confusion_matrix.to_df()` if you just need aggregates. See the ClassificationValidator API, the Image Classification task page (custom validator example), and Val mode export examples. ([docs.ultralytics.com](https://docs.ultralytics.com/reference/data/dataset/))

**Comment by liusse** [comment]
> Thanks! You can log misclassified samples (with file path and true/pred labels) by using a tiny custom `ClassificationDataset` + `ClassificationValidator` that adds `im_file` to the batch and writes a CSV during `val()`; example below.
> 
> import csv
> from pathlib import Path
> import torch
> from ultralytics import YOLO
> from ultralytics.data.dataset import ClassificationDataset
> from ultralytics.models.yolo.classify import ClassificationValidator
> 
> # Include file paths in classification batches
> class DatasetWithPaths(ClassificationDataset):
>     def __getitem__(self, i: int):
>         d = super().__getitem__(i)          # returns {'img': tensor, 'cls': int}
>         d["im_file"] = self.samples[i][0]   # add image file path
>         return d
> 
> # Record misclassifications to CSV
> class MisclsValidator(ClassificationValidator):
>     def build_dataset(self, img_path: str):
>         return DatasetWithPaths(root=img_path, args=self.args, augment=False, prefix=self.args.split)
> 
>     def init_metrics(self, model):
>         super().init_metrics(model)
>         self.mis = []
> 
>     def update_metrics(self, preds: torch.Tensor, batch: dict):
>         super().update_metrics(preds, batch)                    # keep top1/top5 + confusion matrix
>         top1 = preds.argmax(1)
>         probs = preds.softmax(1)
>         for i, (t, p) in enumerate(zip(batch["cls"].view(-1), top1)):
>             if p != t:
>                 self.mis.append((batch["im_file"][i], int(t), int(p), float(probs[i, p])))
> 
>     def finalize_metrics(self):
>         super().finalize_metrics()
>         if self.mis:
>             out = self.save_dir / "misclassified.csv"
>             with open(out, "w", newline="") as f:
>                 w = csv.writer(f)
>                 w.writerow(["file", "true", "pred", "conf", "true_name", "pred_name"])
>                 for fpath, t, p, conf in self.mis:
>                     w.writerow([fpath, t, p, conf, self.names[t], self.names[p]])
> 
> # Run validation with the custom validator
> model = YOLO("yolo11n-cls.pt")
> metrics = model.val(data="path/to/your/val", validator=MisclsValidator)
> Notes and references: `ClassificationDataset.__getitem__` returns only `{'img','cls'}`, so we add `im_file` ourselves; passing a custom validator/dataset is supported as shown in the Image Classification docs; and you can also export the confusion matrix via `results.confusion_matrix.to_df()` if you just need aggregates. See the ClassificationValidator API, the Image Classification task page (custom validator example), and Val mode export examples. ([docs.ultralytics.com](https://docs.ultralytics.com/reference/data/dataset/))

Thank you !
this method is effective.

**Comment by glenn-jocher** [comment]
Great to hear—if you’d also like thumbnails of only the misclassified samples saved during val, run with `plots=True` and override `plot_predictions()` as below.

```python
import torch
from ultralytics.utils.plotting import plot_images
from ultralytics.models.yolo.classify import ClassificationValidator

class MisclsValidator(ClassificationValidator):
    def plot_predictions(self, batch, preds, ni):
        top1 = preds.argmax(1)
        mis = (top1 != batch["cls"].view(-1))
        if mis.any():
            labels = {
                "img": batch["img"][mis],
                "cls": batch["cls"][mis],
                "batch_idx": torch.arange(int(mis.sum()))
            }
            plot_images(labels=labels,
                        fname=self.save_dir / f"val_batch{ni}_mis.jpg",
                        names=self.names,
                        on_plot=self.on_plot)
```

Call with:
```python
metrics = model.val(data="path/to/val", plots=True, validator=MisclsValidator)
```

Method details are in the [ClassificationValidator reference](https://docs.ultralytics.com/reference/models/yolo/classify/val/#plot_predictions), and `plots` is documented under [Val mode arguments](https://docs.ultralytics.com/modes/val/#arguments-for-yolo-model-validation).

**Comment by liusse** [comment]
> on_plot

Good idea! i will try it soon
thanks!!

**Comment by glenn-jocher** [comment]
Sounds good—tip: if you want programmatic access to each saved thumbnail, override `on_plot(name, data)` in your validator to intercept the file path as it’s created; files also land in `metrics.save_dir` after `val()` as a convenience. See the [BaseValidator.on_plot reference](https://docs.ultralytics.com/reference/engine/validator/#on_plot).

```python
from ultralytics.models.yolo.classify import ClassificationValidator

class MisclsValidator(ClassificationValidator):
    def on_plot(self, name, data=None):
        super().on_plot(name, data)  # keep default registration
        print(f"Saved plot: {name}")  # or push to your logger
```
